pallets/token-dealer/src/support.rs

```diff
@@ -0,0 +1,188 @@
+use xcm_executor::traits::{LocationConversion, MatchesFungible, TransactAsset};
+
+pub trait CurrencyIdConversion<CurrencyId> {
+	fn from_asset(asset: &MultiAsset) -> Option<CurrencyId>;
+}
+
+/// The handler for processing cross-chain messages
+pub struct MultiCurrencyAdapter<
+	NativeCurrency,
+	TokenFactory,
+	Matcher,
+	AccountIdConverter,
+	AccountId,
+	CurrencyIdConverter,
+	CurrencyId,
+>(
+	PhantomData<(
+		NativeCurrency,
+		TokenFactory,
+		Matcher,
+		AccountIdConverter,
+		AccountId,
+		CurrencyIdConverter,
+		CurrencyId,
+	)>,
+);
+
+impl<
+		NativeCurrency: Currency<AccountId>,
+		TokenFactory: token_factory::TokenMinter<Ticker, AccountId, NativeCurrency::Balance>,
+		Matcher: MatchesFungible<NativeCurrency::Balance>,
+		AccountIdConverter: LocationConversion<AccountId>,
+		AccountId: sp_std::fmt::Debug,
+		CurrencyIdConverter: CurrencyIdConversion<CurrencyId>,
+	> TransactAsset
+	for MultiCurrencyAdapter<
+		NativeCurrency,
+		TokenFactory,
+		Matcher,
+		AccountIdConverter,
+		AccountId,
+		CurrencyIdConverter,
+		CurrencyId,
+	>
+{
+	fn deposit_asset(asset: &MultiAsset, location: &MultiLocation) -> XcmResult {
+		debug::info!("------------------------------------------------");
+		debug::info!(
+			">>> trying deposit. asset: {:?}, location: {:?}",
+			asset,
+			location
+		);
+		let who = AccountIdConverter::from_location(location).ok_or(())?;
+		debug::info!("who: {:?}", who);
+		let currency = CurrencyIdConverter::from_asset(asset).ok_or(())?;
+		debug::info!("currency_id: {:?}", currency);
+		let amount: NativeCurrency::Balance = Matcher::matches_fungible(&asset).ok_or(())?;
+		debug::info!("amount: {:?}", amount);
+		// match on currency variant
+		if let CurrencyId::Token(token_id) = currency {
+			// mint erc20 token to `who`
+			TokenFactory::mint(token_id, who, amount).map_err(|_| ())?;
+		} else {
+			// native currency transfer via `frame/pallet_balances` is only other variant
+			NativeCurrency::deposit_creating(&who, amount);
+		}
+		debug::info!(">>> successful deposit.");
+		debug::info!("------------------------------------------------");
+		Ok(())
+	}
+
+	fn withdraw_asset(
+		asset: &MultiAsset,
+		location: &MultiLocation,
+	) -> result::Result<MultiAsset, Error> {
+		debug::info!("------------------------------------------------");
+		debug::info!(
+			">>> trying withdraw. asset: {:?}, location: {:?}",
+			asset,
+			location
+		);
+		let who = AccountIdConverter::from_location(location).ok_or(())?;
+		debug::info!("who: {:?}", who);
+		let currency = CurrencyIdConverter::from_asset(asset).ok_or(())?;
+		debug::info!("currency_id: {:?}", currency);
+		let amount: NativeCurrency::Balance = Matcher::matches_fungible(&asset).ok_or(())?;
+		debug::info!("amount: {:?}", amount);
+		// match on currency variant
+		if let CurrencyId::Token(token_id) = currency {
+			// burn erc20 token from `who`
+			TokenFactory::burn(token_id, who, amount).map_err(|_| ())?;
```

Comment:
Shall we log the reason of the error? Just in case we need the withdraw did not take place